internal/tests/cluster/session_cleanup_test.go

```diff
@@ -0,0 +1,513 @@
+	)
+	require.NoError(err)
+	defer p1.Close()
+	require.NotEmpty(t, p1.ListenerAddr())
+
+	// *************
+	// ** Proxy 2 **
+	// *************
+	require.Len(c2.ClusterAddrs(), 1)
+	p2, err := dawdle.NewProxy("tcp", "", c2.ClusterAddrs()[0],
+		dawdle.WithListener(pl2),
+		dawdle.WithRbufSize(512),
+		dawdle.WithWbufSize(512),
+	)
+	require.NoError(err)
+	defer p2.Close()
+	require.NotEmpty(t, p2.ListenerAddr())
+
+	// ************
+	// ** Worker **
+	// ************
+	w1 := worker.NewTestWorker(t, &worker.TestWorkerOpts{
+		WorkerAuthKms:      c1.Config().WorkerAuthKms,
+		InitialControllers: []string{p1.ListenerAddr(), p2.ListenerAddr()},
+		Logger:             logger.Named("w1"),
+	})
+	defer w1.Shutdown()
+
+	time.Sleep(10 * time.Second)
+	expectWorkers(t, c1, w1)
+	expectWorkers(t, c2, w1)
+
+	// Use an independent context for test things that take a context so
+	// that we aren't tied to any timeouts in the controller, etc. This
+	// can interfere with some of the test operations.
+	ctx := context.Background()
+
+	// Connect target
+	client := c1.Client()
+	client.SetToken(c1.Token().Token)
+	tcl := targets.NewClient(client)
+	tgt, err := tcl.Read(ctx, "ttcp_1234567890")
+	require.NoError(err)
+	require.NotNil(tgt)
+
+	// Create test server, update default port on target
+	ts := newTestTcpServer(t, logger)
+	require.NotNil(ts)
+	defer ts.Close()
+	tgt, err = tcl.Update(ctx, tgt.Item.Id, tgt.Item.Version, targets.WithTcpTargetDefaultPort(ts.Port()))
+	require.NoError(err)
+	require.NotNil(tgt)
+
+	// Authorize and connect
+	sess := newTestSession(ctx, t, tcl, "ttcp_1234567890")
+	sConn := sess.Connect(ctx, t, logger)
+
+	// Run initial send/receive test, make sure things are working
+	sConn.TestSendRecvAll(t)
+
+	// Kill connection to first controller, and run test again, should
+	// pass, deferring to other controller.
+	p1.Pause()
+	sConn.TestSendRecvAll(t)
+
+	// Resume first controller, pause second. This one should work too.
+	p1.Resume()
+	p2.Pause()
+	sConn.TestSendRecvAll(t)
+
+	// Kill the first controller connection again. This one should fail
+	// due to lack of any connection.
+	p1.Pause()
+	sConn.TestSendRecvFail(t)
+
+	// Finally resume both, try again. Should behave as per normal.
+	p1.Resume()
+	p2.Resume()
+	time.Sleep(time.Second * 10)                          // Sleep to wait for worker to report back as healthy
+	sess = newTestSession(ctx, t, tcl, "ttcp_1234567890") // re-assign, other connection will close in t.Cleanup()
+	sConn = sess.Connect(ctx, t, logger)
+	sConn.TestSendRecvAll(t)
+}
+
+// testSession represents an authorized session.
+type testSession struct {
+	workerAddr      string
+	transport       *http.Transport
+	tofuToken       string
+	connectionsLeft int32
+}
+
+// newTestSession authorizes a session and creates all of the data
+// necessary to initialize.
+func newTestSession(
+	ctx context.Context,
+	t *testing.T,
+	tcl *targets.Client,
+	targetId string,
+) *testSession {
+	require := require.New(t)
```

Comment:
should we add a call to t.Helper() here and all the testSession receiver functions?

Reply:
This should be addressed now!